chore: Update Argo Workflows to 3.7

aniketpati1121 opened the pull request:
This PR upgrades Argo Workflows to version 3.7 in the manifests. 

Changes include:
- Added RBAC patch files for workflow-controller and workflowtaskresults
- Ensured the workflows have the necessary permissions to run correctly
- Verified hello-world example workflow runs successfully

Please review and let me know if any further adjustments are required. 

fixes #12184

mprahl commented:
@aniketpati1121 thanks for the PR! Please follow the directions here to upgrade Argo Workflows:
https://github.com/kubeflow/pipelines/tree/master/third_party/argo

Also, did you mean to commit those SDK files?

gmfrasca commented:
Hey @aniketpati1121!  We've actually made some updates to the way these Argo upgrades are done; we've added some documentation, but may have missed other places.

That said, this can be done in a couple steps:
1. Update the version specified for the argo-workflows dependency in `go.mod`, then run `go mod tidy`.
1. Update the contents of`third_party/argo/VERSION` to the target version, the run `make update` from the `manifests/kustomize/third-party/argo` directory to update the manifests
1. Finally, look for any outlying references and update them.  specifically, `backend/Dockerfile` and a few docker commands found in the `test` directory also reference it.

We also just added some CI checks for running argo compatibility with KFP, so you can update the values for `argo_version` in `.github/workflows/kfp-samples.yml`, and the list of `argo_version`s in `.github/workflows/e2e-test.yml` to ensure we maintain compatibility with the 'old' version(s) of argo, namely 3.6.7 and 3.5.14.

All of this we should've already added to our documentation such that anyone can pick up this task and run, but that's a different issue :grimacing:

aniketpati1121 commented:
> Hey @aniketpati1121! We've actually made some updates to the way these Argo upgrades are done; we've added some documentation, but may have missed other places.
> 
> That said, this can be done in a couple steps:
> 
> 1. Update the version specified for the argo-workflows dependency in `go.mod`, then run `go mod tidy`.
> 2. Update the contents of`third_party/argo/VERSION` to the target version, the run `make update` from the `manifests/kustomize/third-party/argo` directory to update the manifests
> 3. Finally, look for any outlying references and update them.  specifically, `backend/Dockerfile` and a few docker commands found in the `test` directory also reference it.
> 
> We also just added some CI checks for running argo compatibility with KFP, so you can update the values for `argo_version` in `.github/workflows/kfp-samples.yml`, and the list of `argo_version`s in `.github/workflows/e2e-test.yml` to ensure we maintain compatibility with the 'old' version(s) of argo, namely 3.6.7 and 3.5.14.
> 
> All of this we should've already added to our documentation such that anyone can pick up this task and run, but that's a different issue 😬



Hi @gmfrasca 

I have updated Argo to v3.7.0 across go.mod, go.sum, third_party/argo/VERSION, backend/Dockerfile, and the kfp-samples workflow. 
All necessary changes suggested in the review have been applied, and pre-commit hooks have passed. The PR is now ready for CI re-run and review.

Thanks

rimolive commented:
/ok-to-test

rimolive commented:
@aniketpati1121 I can't say for sure, but probably it's because you did not sign-off your latest commits.

Correct that, and if the test don't trigger again, we'll need the approvers to help.

aniketpati1121 commented:
> @aniketpati1121 I can't say for sure, but probably it's because you did not sign-off your latest commits.
> 
> Correct that, and if the test don't trigger again, we'll need the approvers to help.

Hi @rimolive,  
I have fixed the DCO issue and now all checks have passed successfully .  
The PR is ready for review and approval. Thanks a lot for your guidance!

rimolive commented:
@HumairAK @mprahl Looks like you must approve the workflows run. Please?

aniketpati1121 commented:
Hi, I’ve updated go.mod, go.sum and verified Argo v3.7 upgrade. All DCO and dependency checks have passed. Pending workflows just need approval. Thanks!

gmfrasca commented:
Hey @aniketpati1121 - if you haven't seen it, looks like CI ran but there's some build failures.  Seems like it's because Argo updated the validate.ValidateWorkflow method signature to take in an additional parameter, though there may be additional updates to be made as well.  

I think once you make the updates to any of those calls there, and are able to get`make images_all` to complete (or just run the `go build` commands directly), most of the CI should run and pass.

gmfrasca commented on manifests/kustomize/third-party/argo/installs/cluster/kustomization.yaml:
```diff
@@ -1,12 +1,9 @@
 resources:
-- https://github.com/argoproj/argo-workflows/manifests/base/crds?ref=v3.6.7
-- https://github.com/argoproj/argo-workflows/manifests/cluster-install/workflow-controller-rbac?ref=v3.6.7
+- https://github.com/argoproj/argo-workflows/manifests/base/crds?ref=v3.7.0
+- https://github.com/argoproj/argo-workflows/manifests/cluster-install/workflow-controller-rbac?ref=v3.7.0
```
looks like argo shuffled their manifests a bit in v3.7.0, which is why `kustomize build` fails to find these remote refs.  I think this should fix it:  

```suggestion
- https://github.com/argoproj/argo-workflows/manifests/base/crds/minimal?ref=v3.7.0
- https://github.com/argoproj/argo-workflows/manifests/cluster-install-no-crds/workflow-controller-rbac?ref=v3.7.0

```

aniketpati1121 commented:
Closing this in favor of new clean PR
